models/User.go

```diff
@@ -1764,10 +1797,20 @@ func PerformOrderFillTransaction(ask *Ask, bid *Bid, stockTradePrice uint64, sto
 	// save old cash for rolling back
 	askingUserOldCash := askingUser.Cash
 	biddingUserOldCash := biddingUser.Cash
+	askingUserOldReservedCash := askingUser.ReservedCash
+	biddingUserOldReservedCash := biddingUser.ReservedCash
 
 	//calculate user's updated cash
 	askingUser.Cash += uint64(stockTradeQty) * stockTradePrice
-	biddingUser.Cash = uint64(cashLeft)
+
+	// bidding user's cash will first be subtracted from reserved cash and then if required from bidding user's cash
+	cashDifference := int64(biddingUser.ReservedCash) - total
+	if cashDifference >= 0 {
+		biddingUser.ReservedCash -= uint64(total)
+	} else {
+		biddingUser.ReservedCash = 0                                        // reserved cash gets all used
+		biddingUser.Cash -= uint64(total - int64(biddingUser.ReservedCash)) // remaining cash is used completely, money taken from cash
```

Comment:
This line should not be there. It was correct the way it was before (i.e) `biddingUser.Cash = uint64(cashLeft)`.

All you need to do is update `biddingUser.ReservedCash`. And for that it would be better if you used 

```go
biddingUser.ReservedCash = math.Max(0, biddingUser.ReservedCash - total)
```

This is simpler than using if else statements.

Reply:
```
It was correct the way it was before (i.e) `biddingUser.Cash = uint64(cashLeft)`.
@gauthamk97 
```
Say if user purchased stocks, but gets it for less, then only reserved cash would be deducted, user cash won't change,
Doing your way would cause user cash to increase also which was required earlier but now not required as there is reserved cash for that.

<hr>

Better way would be

``` biddingUser.ReservedCash = math.Max(0, biddingUser.ReservedCash - total) ```
``` biddingUser.Cash += math.Min(0, biddingUser.ReservedCash - total) ```